securedrop/tests/test_encryption.py

```diff
@@ -0,0 +1,355 @@
+        assert decrypted_reply == journalist_reply
+
+        # And source2 is NOT able to decrypt the reply
+        with pytest.raises(GpgDecryptError):
+            encryption_mgr.decrypt_journalist_reply(
+                for_source_user=source_user2, ciphertext_in=encrypted_reply
+            )
+
+        # Amd the reply can't be decrypted without providing the source1's gpg secret
+        result = encryption_mgr._gpg.decrypt(
+            # For GPG 2.1+, a non-null passphrase _must_ be passed to decrypt()
+            encrypted_reply,
+            passphrase="test 123",
+        )
+        assert not result.ok
+
+        # And the journalist is able to decrypt their reply
+        with import_journalist_private_key(encryption_mgr):
+            decrypted_reply_for_journalist = encryption_mgr._gpg.decrypt(
+                # For GPG 2.1+, a non-null passphrase _must_ be passed to decrypt()
+                encrypted_reply,
+                passphrase="test 123",
+            ).data
+        assert decrypted_reply_for_journalist.decode() == journalist_reply
+
+    def test_encrypt_fails(self, setup_journalist_key_and_gpg_folder, tmp_path):
+        # Given an encryption manager
+        journalist_key_fingerprint, gpg_key_dir = setup_journalist_key_and_gpg_folder
+        encryption_mgr = EncryptionManager(
+            gpg_key_dir=gpg_key_dir, journalist_key_fingerprint=journalist_key_fingerprint
+        )
+
+        # When trying to encrypt some data without providing any recipient
+        # It fails and the right exception is raised
+        with pytest.raises(GpgEncryptError) as exc:
+            encryption_mgr._encrypt(
+                using_keys_with_fingerprints=[],
+                plaintext_in="test",
+                ciphertext_path_out=tmp_path / "encrypt_fails",
+            )
+        assert "no terminal at all requested" in str(exc)
+
+
+def _parse_gpg_date_string(date_string: str) -> datetime:
+    """Parse a date string returned from `gpg --with-colons --list-keys` into a datetime.
+
+    The format of the date strings is complicated; see gnupg doc/DETAILS for a
+    full explanation.
+
+    Key details:
+    - The creation date of the key is given in UTC.
+    - the date is usually printed in seconds since epoch, however, we are
+    migrating to an ISO 8601 format (e.g. "19660205T091500"). A simple
+    way to detect the new format is to scan for the 'T'.
+    """
+    if "T" in date_string:
+        dt = datetime.strptime(date_string, "%Y%m%dT%H%M%S")
+    else:
+        dt = datetime.utcfromtimestamp(int(date_string))
+    return dt
+
+
+@contextmanager
+def import_journalist_private_key(
+    encryption_mgr: EncryptionManager,
+) -> typing.Generator[None, None, None]:
+    """Import the journalist secret key so the encryption_mgr can decrypt data for the journalist.
+
+    The journalist secret key is removed at the end of this context manager in order to not impact
+    other decryption-related tests.
+    """
```

Comment:
`import_journalist_private_key()` can be used when the journalist secret key needs to be available for a specific test.